### PR TITLE
8329603: G1: Merge G1BlockOffsetTablePart into G1BlockOffsetTable

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch merges `G1BlockOffsetTablePart` into `G1BlockOffsetTable`. The previous fields `_reserved` and `_offset_base` of `G1BlockOffsetTable` are marked as `static` so that they can be shared by BOTs of all the heap regions.

The tests `make test-tier1_gc` passed locally. Thanks for taking the time to review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329603](https://bugs.openjdk.org/browse/JDK-8329603): G1: Merge G1BlockOffsetTablePart into G1BlockOffsetTable (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [a8a121bf](https://git.openjdk.org/jdk/pull/18634/files/a8a121bf9aa0ebc822dd22f098ff4969498773f0)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18634/head:pull/18634` \
`$ git checkout pull/18634`

Update a local copy of the PR: \
`$ git checkout pull/18634` \
`$ git pull https://git.openjdk.org/jdk.git pull/18634/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18634`

View PR using the GUI difftool: \
`$ git pr show -t 18634`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18634.diff">https://git.openjdk.org/jdk/pull/18634.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18634#issuecomment-2037726691)

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp]: I marked the fields `_offset_base` and `_reserved` as `non-static`, removed the field `_hr` and shared a simple/unified BOT for all the heap regions. Since few methods need the heap region, I used the heap region as method parameter in such methods.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:134]: I feel this doesn't belong to BOT. Can probably be dealt with in another ticket.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp:166]: Some indentation issues.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:43]: I think it is good to delete this line and the blank line. What do you think about it?

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:134]: That sounds reasonable. (Should not be done in this PR though.)

(My experience with BOT is that they are almost never corrupted, so doing only checking-after-each-write is enough, sth like `ObjectStartArray::verify_for_block` -- there is possibly little value in verifying BOT in `HeapRegion::verify`. I wonder what others' opinions are.)

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:43]: Removing it is fine, IMO.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp]: Fixed. I tried to find other indentation issues except this one, but can't find now. Please point it out if you found.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp]: Removed.

File: src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1BlockOffsetTable.hpp:131]: > I feel this doesn't belong to BOT. Can probably be dealt with in another ticket.

Filed https://bugs.openjdk.org/browse/JDK-8329767 to follow up.

> (My experience with BOT is that they are almost never corrupted, so doing only checking-after-each-write is enough, sth like `ObjectStartArray::verify_for_block` -- there is possibly little value in verifying BOT in `HeapRegion::verify`. I wonder what others' opinions are.)

The verifications of Serial and Parallel BOTs are similar. I agree that G1 should be adjusted. Filed https://bugs.openjdk.org/browse/JDK-8329771 to follow up.

**Comment by lgxbslgx** [comment]
@albertnetymk Thanks for your review. Waiting for another review.

File: src/hotspot/share/gc/g1/g1HeapRegion.inline.hpp
Comment (walulyai) [inline src/hotspot/share/gc/g1/g1HeapRegion.inline.hpp:107]: Can we move the assert `assert(addr >= _hr->bottom() && addr < _hr->top(), "invalid address");` here?

File: src/hotspot/share/gc/g1/g1HeapRegion.hpp
Comment (walulyai) [inline src/hotspot/share/gc/g1/g1HeapRegion.hpp:77]: I suppose there is no longer a reason for this to be part of the region. Any downside to referring to it using `g1h->bot()`?

**Review by walulyai — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1HeapRegion.hpp
Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1HeapRegion.hpp:77]: That would require calling `G1CollectedHeap::heap()` every time bot is needed. Actually, maybe some methods, e.g. `update_bot_for_block` should belong to heap, instead of heap-region.

(Either way, I feel this decision can/should be made in its own ticket.)

File: src/hotspot/share/gc/g1/g1HeapRegion.inline.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1HeapRegion.inline.hpp:107]: Moved.

File: src/hotspot/share/gc/g1/g1HeapRegion.hpp
Comment (lgxbslgx) [reply src/hotspot/share/gc/g1/g1HeapRegion.hpp:77]: > (Either way, I feel this decision can/should be made in its own ticket.)

I agree. This patch should focus on merging.

**Review by walulyai — APPROVED** [review]

**Pull request closed without merging**
